lib/sidekiq/batch.rb

```diff
@@ -38,62 +37,148 @@ def on(event, callback, options = {})
+
+        @ready_to_queue
+      ensure
+        Thread.current[:bid_data] = bid_data
+      end
+    end
+
+    def increment_job_queue(jid)
+      @ready_to_queue << jid
     end
 
     private
 
     def persist_bid_attr(attribute, value)
       Sidekiq.redis do |r|
         r.multi do
-          r.hset("BID-#{bid}", attribute, value)
-          r.expire("BID-#{bid}", BID_EXPIRE_TTL)
+          r.hset(@bidkey, attribute, value)
+          r.expire(@bidkey, BID_EXPIRE_TTL)
         end
       end
     end
 
     class << self
       def process_failed_job(bid, jid)
-        to_process = Sidekiq.redis do |r|
+        _, pending, failed, children, complete = Sidekiq.redis do |r|
           r.multi do
             r.sadd("BID-#{bid}-failed", jid)
+
+            r.hincrby("BID-#{bid}", "pending", 0)
             r.scard("BID-#{bid}-failed")
-            r.hget("BID-#{bid}", 'to_process')
+            r.hincrby("BID-#{bid}", "children", 0)
+            r.scard("BID-#{bid}-complete")
+
             r.expire("BID-#{bid}-failed", BID_EXPIRE_TTL)
           end
         end
-        if to_process[2].to_i == to_process[1].to_i
-          Callback.call_if_needed(:complete, bid)
-        end
+
+        enqueue_callback(:complete, bid) if pending.to_i == failed.to_i && children == complete
       end
 
-      def process_successful_job(bid)
-        out = Sidekiq.redis do |r|
+      def process_successful_job(bid, jid)
+        failed, pending, children, complete, success, total, parent_bid = Sidekiq.redis do |r|
           r.multi do
-            r.hincrby("BID-#{bid}", 'to_process', -1)
             r.scard("BID-#{bid}-failed")
-            r.hincrby("BID-#{bid}", 'pending', -1)
+            r.hincrby("BID-#{bid}", "pending", -1)
+            r.hincrby("BID-#{bid}", "children", 0)
+            r.scard("BID-#{bid}-complete")
+            r.scard("BID-#{bid}-success")
+            r.hget("BID-#{bid}", "total")
+            r.hget("BID-#{bid}", "parent_bid")
+
+            r.srem("BID-#{bid}-failed", jid)
+            r.srem("BID-#{bid}-jids", jid)
             r.expire("BID-#{bid}", BID_EXPIRE_TTL)
           end
         end
 
         puts "processed process_successful_job"
-        Callback.call_if_needed(:complete, bid) if out[1].to_i == out[0].to_i
-        Callback.call_if_needed(:success, bid) if out[0].to_i.zero?
+
+        enqueue_callback(:complete, bid) if pending.to_i == failed.to_i && children == complete
+        enqueue_callback(:success, bid) if pending.to_i.zero? && children == success
+      end
+
+      def enqueue_callback(event, bid)
+        needed, _, callback, opts, queue, parent_bid = Sidekiq.redis do |r|
+          r.multi do
+            r.hget("BID-#{bid}", event)
+            r.hset("BID-#{bid}", event, true)
+            r.hget("BID-#{bid}", "callback_#{event}")
+            r.hget("BID-#{bid}", "callback_#{event}_opts")
+            r.hget("BID-#{bid}", "callback_queue")
+            r.hget("BID-#{bid}", "parent_bid")
```

Comment:
No, because this function can also be called from the callback.rb with the bid of the parent.

Maybe this file is not the correct location for the function then.